### ISSUE TITLE
Unify the ast and token worlds

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#1440]
[leoAst.py](https://github.com/leo-editor/leo-editor/blob/fstrings/leo/core/leoAst.py) unifies python's token-oriented and ast-oriented worlds. This project is intended to be a major contribution to python's tool set.

**Overview**

The **TokenOrderGenerator** (TOG) class in leoAst.py creates the following data:

- A **children** array from each ast node to its children. Order matters!
- A **parent** link from each ast.node to its parent.
- Two-way links between tokens in the **token list**, a list of Token objects, and the [ast](https://docs.python.org/3/library/ast.html) nodes in the **parse tree**:
  - For each token, **token.node** contains the ast.node "responsible" for the token.
  - For each ast node, **node.first_i** and **node.last_i** are indices into the token list.
    These indices give the range of tokens that can be said to be "generated" by the ast node.

These links promise to collapse the complexity of any code that changes text, including the [asttokens](https://pypi.org/project/asttokens/), [fstringify](https://pypi.org/project/fstringify/), and [black](https://pypi.org/project/black/) projects.

This project is a general solution to frequently-asked questions such as [this](https://stackoverflow.com/questions/16748029). It fills gaps in python's ast module, as discussed in [this python issue](https://bugs.python.org/issue33337).

**Project details**

Leo's [fstrings branch](https://github.com/leo-editor/leo-editor/tree/fstrings) contains this work, as well as work on Leo's fstringifier command. This branch will be merged into [devel](https://github.com/leo-editor/leo-editor/tree/devel) from time to time.

leoAst.py is completely independent of [Leo](http://leoeditor.com/) itself. leoAst.py contains a complete suite of unit tests. Those unit tests completely cover the most important classes of the file.

Work is in progress on distributing leoAst.py as a separate project. In the meantime, you may find the latest version of leoAst.py [here](https://github.com/leo-editor/leo-editor/blob/fstrings/leo/core/leoAst.py).

**Status**

*The project is a complete success*.  The code in leoAst.py is simpler, easier to understand, more flexible, more robust and faster than the corresponding code in the asttokens, fstringify and black projects. See the following section for more details.

leoAst.py contains an entirely new implementation of the [fstringify](https://pypi.org/project/fstringify/) tool. This new fstringify is simpler, faster, more complete, more flexible and more reliable than fstringify itself.

leoAst.py also contains an entirely new implementation of [black](https://pypi.org/project/black/). This code is under development. Basing the code on the TOG class will surely make the tool simpler, more flexible and faster than black itself.

**Figures of merit**

**Simplicity**: The code (all of it) is the simplest thing that could possibly work. This is in stark contrast to the generators used in the [asttokens](https://pypi.org/project/asttokens/), [fstringify](https://pypi.org/project/fstringify/), and [black](https://pypi.org/project/black/) projects.

**Flexibility**: Flexibility comes from simplicity, not special cases. The code contains no hacks of any kind. Again, this is in stark contrast to the generators used by asttokens, fstringify and black.

**Speed**: The TOG creates two-way links between tokens and ast nodes in roughly the time taken by python's tokenize.tokenize and ast.parse library methods. This is substantially faster than the  [asttokens](https://pypi.org/project/asttokens/), [fstringify](https://pypi.org/project/fstringify/), and [black](https://pypi.org/project/black/) tools. The TOT class traverses trees annotated with parent/child links even more quickly.

**Memory**: The TOG class makes no significant demands on python's resources. Generators add nothing to python's call stack. TOG.node_stack is the only variable length data. This stack resides in python's heap, so its length is unimportant. In the worst case, it might contain a few thousand entries. The TOT class uses no variable-length data whatever.

**Innovation**: The code's speed, simplicity, robustness and flexibility is the result of months of work.  For more details, see the Project's history in [this issue's third comment](https://github.com/leo-editor/leo-editor/issues/1440#issuecomment-574145510).

**Testing and inspection**: The AstDumper class shows all the links in various formats. Strong unit tests cover all parts of the TOG and TOT classes.

**To do**

Infrastructure:
- [x] Create Token, Tokenizer, TokenOrderGenerator and TokenOrderTraverser classes.
- [x] Create a set of dumping tools.
- [x] Create nearest_common_ancestor and tokens_for_node functions.

Testing:
- [x] tokenizer.check_results verifies token round-tripping.
- [x] tog.sync_tokens is an ever-present unit test that verifies that the TOG visits nodes in token order.
- [x] Create unit tests that cover the TOG, TOT and Fstringify classes.
- [x] 100% coverage of all non-testing code.
- [x] Beautify all files.
- [x] Fstringify all files.

Real world tools:
- [x] Create the Fstringify class.
- [x] Create fstringify-files and diff-fstringify-files commands.
- [x] Rewrite Leo's beautify-files and diff-beautify-files commands using the Orange class.
- [x] Complete the Orange class.
- [x] **Won't do**: Create orange.py, for stand-alone operation.
    - Black if good for non-Leo files.
    - Leo's beautifier commands are needed for *all* Leo files.

**Comment by edreamleo** [comment]
**Theory of Operation**

To do.

**Comment by edreamleo** [comment]
**Project history**

EKR's posts record the many Aha's there have been along the way:

Part 1: Initial steps.

- Nov. 10, 2019: [ENB: The unification of the token and ast worlds](https://groups.google.com/forum/#!topic/leo-editor/FZYJmbtRBWs) contains the initial goal.
- Nov 13, 2019: [ENB: flattening the TokenOrderTraverser class](https://groups.google.com/d/msg/leo-editor/2vXq5F-rsdQ/vKYHUGsqAwAJ).
- Nov 15, 2019: [ENB: Two Ahas re unifying the token and parse-tree worlds](https://groups.google.com/d/msg/leo-editor/6JfTBs7i73o/CWZVHXV_AwAJ).
- Nov 16, 2019: [ENB: still a juicy project](https://groups.google.com/d/msg/leo-editor/malKiU__CUY/0WJJcIHXAwAJ).
- Nov 16, 2019: [ENB: The intuition wars are over](https://groups.google.com/d/msg/leo-editor/cw77W5XEgDw/gDrN6nrkAwAJ). TOG clearly can work in linear time.
- Nov 18, 2019: [Success! High-level overview and status report](https://groups.google.com/d/msg/leo-editor/PsB8dTr2Wac/Yh49dHh6BAAJ).
- Nov 19, 2019: [The Linker class has collapsed!](https://groups.google.com/d/msg/leo-editor/n0nRwcVHFQo/87Jx_EwDBAAJ). Eventually the linker class completely disappeared...
- Nov 22, 2019: [Brief status report; long ENB re Ahas and code](https://groups.google.com/d/msg/leo-editor/x3F8q_Gxvws/jP_13dGOAAAJ). At this point I knew that the do_If visitor could not distinguish between "else if" and "elif".

Part 2: A stupendous collapse in complexity.

Nov 23, 2019 was the most momentous day in this project's history. [ENB: Wow: the entire project will be nothing but two parallel generators](https://groups.google.com/d/msg/leo-editor/_tvE6lwDKHg/mMQ0GufMAAAJ):

**First Aha**: "Neither the results list nor the Linker class is needed! This is a stunning collapse in complexity....Linker.set_links only sees significant tokens.  It could migrate to tog.put [which later became tog.sync_token]." This was followed by a lengthy discussion of details.

**Second Aha**: "That was the end of a long day's work.  However, as I was talking to Rebecca about the success, all the pieces fell into place.

"I'll never forget it.  I was brushing my teeth and saying wow.  A pause.  Another wow.  Another pause.  Another wow. I realized that the "if-list" should be a generator, that a token generator could yield all significant tokens, and that the tree and token generators would work in parallel."

"I explained it to Rebecca as follows. I held both arms over my head.  My right hand represented the tree generator.  It moved back and forth as it moved from top to bottom, simulating traversing the tree.  My left hand represented the token generator.  It moved straight down, because it was just moving from one token to the next in the token list. Crucially, the height of the two hands always matched. The generators work in parallel."

**Third Aha**: Later that day, I wrote [this](https://groups.google.com/d/msg/leo-editor/_tvE6lwDKHg/ph13Da4iAQAJ):

"Doh. The so-called generators are nothing but indices into the token list. There is no need for helper lists! peek and advance need only maintain an integer index into the token list.  This index monotonically advances, guaranteeing O(N) performance."

Part 3: Final details:

- Nov 27, 2019: [ENB: The last puzzle](https://groups.google.com/d/msg/leo-editor/FQRbeq5HJdI/RrIJJu4OAgAJ). At this point I was still trying to use the parse tree as the primary data.
- Dec 5, 2019: [The unification of tokens and parse trees is complete].
- Dec 5, 2019: [Status report and a big Aha](https://groups.google.com/d/msg/leo-editor/wFM8TyIN-9s/Cx0ZcX59BAAJ):

"A giant Aha: We can determine which 'string' tokens are concatenated just by looking at the token list!!! Indeed, 'string' tokens are concatenated if and only if there are no significant tokens (including parens) between them!"

**Comment by edreamleo** [comment]
I am going to close this issue. It is code complete. The Orange and Fstringify classes are the basis of Leo's beautify-files and fstringify-files command. As such, they are substantial illustrations of the TOG and TOT classes.
